design/rest-api-guideline.md

```diff
@@ -0,0 +1,388 @@
+# REST API Guideline
+
+* **Status**: Draft
+* **Author**: [pedroigor](https://github.com/pedroigor), [abstractj](https://github.com/abstractj), [ssilvert](https://github.com/ssilvert)
+* **JIRA**: [KEYCLOAK-9344](https://issues.jboss.org/browse/KEYCLOAK-9344)
+
+## Abstract
+
+The Keycloak REST API Guideline provides a set of design principles and practices that should be considered by developers when designing, implementing and exposing a RESTful API. To provide the best experience for developers consuming Keycloak REST APIs, it's important API developers to follow to the principles and practices defined herein, making these APIs intuitive and easy to use.
+
+The goals of the guideline are:
+
+* Define the [overall principles](#principles) that should be taken into account when designing REST APIs for Keycloak
+
+* Define common practices and patterns for all Keycloak REST APIs such as:
+    * [Filtering](#filtering)
+    * [Pagination](#pagination)
+    * [Error Handling](#error-handling)
+    * [Documentation](#documentation)
+
+* Define common practices for [documentation](#documentation)
+
+Although the guideline is based on the REST architectural style, it does not try to strictly adhere to RESTful constraints. The principles covered by the guideline are based on practical experiences in the development of REST APIs in Keycloak while still conforming with most of the constraints of REST architectures.
+
+**Table of Contents**
+
+
+  - [Principles](#principles)
```

Comment:
We should also consider a way to limit what fields are returned, which is especially relevant for larger objects. In the future I can see us having a single realm json config that contains all config for the realm, instead of separate child resources for user storage providers, etc. Then rely on PATCH to add/remove to the realm config.

Reply:
The filtering part should be covered by the filtering section so that you can fetch only the fields you want (user storage providers) and then use PATCH to update only those fields.

Reply:
We may also consider taking similar approach to GitHub: by default, return only some limited set of fields (basic info); if you need more, you have to specify those extra fields in the filter.

Reply:
+1